### PR TITLE
Data Migration Script for users_settings and users_notification_settings

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR creates two DataUpdate scripts for migrating relevant data from the `users` table to the `users_settings` and `users_notification_settings` tables. There are also associated specs for both scripts.

🛑  &nbsp; NOTE: [per this comment](https://github.com/forem/forem/pull/13390#issuecomment-823983389), we shall implement the sync code for `users_settings` before running these migration scripts. So this PR will remain in Draft until then. But please still review 🙏🏾 

## QA Instructions, Screenshots, Recordings

In order to temporarily run this script, you can simply paste the contents in `rails dbconsole`.

## Added tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [X] This change does not need to be communicated, and this is why not: _this is step 3 in a multi-step RFC_

## [optional] What gif best describes this PR or how it makes you feel?

Thunddercats

### CODE REVIEW

**Comment by Ridhwana** [comment]
Steps to test the script (this is more specifically to test the migration of profile fields):

1. Go to http://localhost:3000/settings and populate the relevant profile fields for your user  (i.e. `brand_color_1`, `brand_color_2` and `display_email_on_profile`. 
(@msarit `display_email_on_profile` still needs to be added to this script)

2. Thereafter, go to your `rails dbconsole` and copy and paste either all the contents or a portion of the contents that you want to run in the console. If you are pasting a portion of the contents, remember to use a `;` at the end for execution. 

3. After pasting the entire script, you can see if the  updates were made to your db by doing the following 
- Run `Select * from users_settings` on the next line whilst in the `dbconsole` to view the updated user settings table, or
-  In the case where you are more comfortable with the active record and the rails console you can open another console  with `bundle exec rails c` and get the user_settings table by running  `Users::Setting.all`. 

If the sql query was successful you should see `brand_color_1`, `brand_color_2` populated as per the Profile Fields table.

If you con't want to commit the changes, and just test that the script executes without errors use `ROLLBACK` instead of `COMMIT`. Also, if at any point you want to delete the contents of the `users_settings` table and start afresh you can run `Users::Setting.destroy_all`


Just a note that each time that we execute the script it makes new rows in the database. We want to ensure that if the script fails midway through and we have to re-run the script that we don't end up with duplicate entries.

Hope this helps @msarit :) Feel free to ping with follow up questions.

File: user_settings.sql [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        display_announcements = EXCLUDED.display_announcements,
+        display_sponsors = EXCLUDED.display_sponsors,
+        feed_mark_canonical = EXCLUDED.feed_mark_canonical,
+        feed_referential_link = EXCLUDED.feed_referential_link,
+        feed_url = EXCLUDED.feed_url,
+        experience_level = EXCLUDED.experience_level,
+        inbox_guidelines = EXCLUDED.inbox_guidelines,
+        updated_at = EXCLUDED.updated_at,
+        brand_color1 = EXCLUDED.brand_color1,
+        brand_color2 = EXCLUDED.brand_color2;
```
Comment (Ridhwana) [inline user_settings.sql:94]: @msarit is this a mistaken duplication?

File: user_settings.sql [hunk 2]
```diff
@@ -0,0 +1,94 @@
+-- This is just a test (in draft) file so that Arit and I can work together using revision control
+-- We run the contents of this file in the rails dbconsole
+-- It will not be committed to the codebase, it will be moved into a data update script and this file will be deleted :)
+
+BEGIN TRANSACTION;
+WITH settings_data AS (
+  SELECT
+    users.id AS user_id,
+    CASE WHEN config_font='default' THEN 0
```
Comment (Ridhwana) [inline user_settings.sql:9]: **nitpick, non-blocking:** can we alphabetize the column names just to make it easier to read and find the names?

File: user_settings.sql [hunk 3]
```diff
@@ -0,0 +1,94 @@
+-- This is just a test (in draft) file so that Arit and I can work together using revision control
+-- We run the contents of this file in the rails dbconsole
+-- It will not be committed to the codebase, it will be moved into a data update script and this file will be deleted :)
+
+BEGIN TRANSACTION;
+WITH settings_data AS (
+  SELECT
```
Comment (Ridhwana) [inline user_settings.sql:7]: I did a cross check of these columns against the users_settings table and noticed that `permit_adjacent_sponsors` is missing. Can we add that please?

File: db/migrate/20210419171746_add_unique_user_id_index_to_users_settings.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class AddUniqueUserIdIndexToUsersSettings < ActiveRecord::Migration[6.1]
+  disable_ddl_transaction!
+
+  def change
```
Comment (Ridhwana) [inline db/migrate/20210419171746_add_unique_user_id_index_to_users_settings.rb:4]: Can we not just remove and add the new index with the unique constraint in one migration file instead of two? 

```
def change:
   remove_index ...
   add_index ...
end
```

**Review by Ridhwana — COMMENTED** [review]
Nice work! Left some comments on this version 😊

File: user_settings.sql (hunk 1 above)
Comment (msarit) [reply user_settings.sql:94]: Oops! 🙈

File: user_settings.sql (hunk 2 above)
Comment (msarit) [reply user_settings.sql:9]: Yes will do 👍🏾

File: user_settings.sql (hunk 3 above)
Comment (msarit) [reply user_settings.sql:7]: Good catch thank you

File: db/migrate/20210419171746_add_unique_user_id_index_to_users_settings.rb (hunk 4 above)
Comment (msarit) [reply db/migrate/20210419171746_add_unique_user_id_index_to_users_settings.rb:4]: Rhymes and I attempted this when working on the SQL statement yesterday and it failed 🤷🏾‍♀️  But when we ran the migrations separately, they were successful.

Comment (rhymes) [reply db/migrate/20210419171746_add_unique_user_id_index_to_users_settings.rb:4]: Yep! @Ridhwana `strong_migrations` was complaining about something (I don't remember exactly :D). Since they are run in quick succession, I don't think it matters that much. We need it to be a unique index so that `ON CONFLICT DO UPDATE` can work

**Comment by Ridhwana** [comment]
So I wanted to drop some rough notes on what I've been thinking about and possibly source some thoughts from people who've done something similar.

**What do we want to achieve and what are some of the considerations:**
- We want to migrate the necessary data from the user table to the users setting table
- We want to migrate the necessary data from the user table to the users notifications setting table 

Working on the assumption that this DUS _may fail_ because we have lots of User data that has evolved over time.  
- If the DUS fails halfway through we want to be able to re-run it without re-populating the existing data
- If the DUS fails and thereafter a user updates an existing row in the User table that as already been exported to the User Settings table, then we need to go back to update the related fields on the User Settings table as well. 

**What happens after we complete both scripts?**

_Option 1:_
- Deploy the data script but only run it on a smaller community (the benhalpern one maybe) to test it with production data. One way to stop it running on DEV nesting is i guess using an statement in the DUS that forces it to only run on certain communities (maybe with a Feature Flag)
- We'd then also need to be able to re-run the script to create and update data for the portion of data between testing the script and syncing the data in realtime.

_Option 2 (this option is outlined in the RFC):_

- Deploy the DUS and data sync code at the same time
- There could possibly be some rows that are updated whilst the DUS is running, how do we handle those cases, have we accounted for that in our script?
- If the script fails halfway, how do we ensure that we can run the DUS again once we fix it, taking into consideration that we've also started syncing data.

Other Thoughts:
- Do we want to pull in the original `created_at` and `updated_at` values from the `User` table or do we want to set new ones for when the migration takes place? 
Maybe we can leverage these timestamps to migrate the data between the first migration and sync in Option 1 or to determine if we have newly updated rows that we missed in Option 2.
- Have we checked some sample data to make sure that the `null: false` fields have values in production as the User table has probably evolved with time. Have we catered for this in our script?

**Comment by Ridhwana** [comment]
I just basically [wrote out some of the things](https://github.com/forem/forem/pull/13390#issuecomment-823346536) I've been thinking about and @msarit and I have been chatting about, whilst we're working on this Data Update Script. @rhymes and @citizen428, wondering if y'all have any comments on it or considerations that you think we need to account for?

**Comment by msarit** [comment]
> Do we want to pull in the original created_at and updated_at values from the User table or do we want to set new ones for when the migration takes place?

I too had this question while building the scripts. As I understand it, our use of `users.created_at` and `users.updated_at` in both scripts above refer to the datetime the `User` was created, and when last the `User` updated their settings, respectively.

But, for `Users::Setting` and `User::NotificationSetting` entries, don't we want their `created_at` values to actually be the record's (and not the user's)?

I hope I'm being clear 😅 . I see the usefulness of having the `users.updated_at` value, as Ridhwana mentioned.

---

> Have we checked some sample data to make sure that the null: false fields have values in production as the User table has probably evolved with time. Have we catered for this in our script?

While working on this with Rhymes, we discovered that some settings that have `null: false` defined on them may currently be `null` in the database. For these, I utilized the `COALESCE` function, which returns the first non-null argument. Wherever I used this function in the scripts above, I passed in the default value for that setting as the second argument.

As an example, for the `users.email_digest_periodic` setting, the default value is `false`. So in the `coalesce` function, I pass in `false` as the alternate value if that setting comes back `null`:
```sql
COALESCE(email_digest_periodic, false)
```

File: db/schema.rb [hunk 5]
```diff
@@ -1376,7 +1376,6 @@
     t.index ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true
     t.index ["twitter_username"], name: "index_users_on_twitter_username", unique: true
     t.index ["username"], name: "index_users_on_username", unique: true
-    t.check_constraint "username IS NOT NULL", name: "users_username_not_null"
```
Comment (Ridhwana) [inline db/schema.rb:1379]: @msarit is this a related schema change?

File: user_settings.sql [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
```
Comment (Ridhwana) [inline user_settings.sql:7]: @msarit based on this comment:
> While working on this with Rhymes, we discovered that some settings that have null: false defined on them may currently be null in the database. For these, I utilized the COALESCE function, which returns the first non-null argument. Wherever I used this function in the scripts above, I passed in the default value for that setting as the second argument.

Have we confirmed that it was only needed on `display_email_on_profile` in the `users_settings` table? Are there any others that require us to use the COALESCE? Should we use this function for all columns that are `null: false`?

**Comment by Ridhwana** [comment]
@msarit in prep for our pairing session today, I  see the next steps of the PR as:
- [ ] Double check the null values for `users_settings` and `users_notification_settings`
- [ ] Test script behaviour if we have to re-run it multiple times and amend as necessary
- [ ] Move script to the DUS
- [ ] Write tests for the DUS

**Comment by rhymes** [comment]
@Ridhwana 

the more I read your proposal, the more I think we should do something similar to the profile migration, in which the code to sync is put in place before the migration script. This is the scenario I imagine:

1. the sync code is deployed
2. users objects have `.save` called upon them
3. this saves the data in `users` and syncs with the new tables using this migration script
4. we deploy a data update script to sync all remaining users
5. we then start migrating the destination columns from `users` to `users_settings`, so that when a profile is saved, the data ends in the latter table

See my inline comments:

> Working on the assumption that this DUS _may fail_ because we have lots of User data that has evolved over time.
> 
>     * If the DUS fails halfway through we want to be able to re-run it without re-populating the existing data

This is why we need the `ON CONFLICT DO UPDATE` in the current script

>     * If the DUS fails and thereafter a user updates an existing row in the User table that as already been exported to the User Settings table, then we need to go back to update the related fields on the User Settings table as well.

This will happen. DEV has 600k users, we can't assume nobody will save their profile while we run a script: the script may file, the script may take a while. We could run it periodically with a scheduled worker, to make sure that, for example, once every hour we sync the two tables with what's in the `users` table

> **What happens after we complete both scripts?**
> 
> _Option 1:_
> 
>     * Deploy the data script but only run it on a smaller community (the benhalpern one maybe) to test it with production data. One way to stop it running on DEV nesting is i guess using an statement in the DUS that forces it to only run on certain communities (maybe with a Feature Flag)

This will need to verify the correctness of the script and data but it won't prepare us for the possibility that there's incosistent data on DEV. We could run it on DEV and then write a SQL query to check if the data has been migrated correctly. Even if the data has been migrated correctly.

>     * We'd then also need to be able to re-run the script to create and update data for the portion of data between testing the script and syncing the data in realtime.

We can do that with a scheduled worker

> _Option 2 (this option is outlined in the RFC):_
> 
>     * Deploy the DUS and data sync code at the same time
> 
>     * There could possibly be some rows that are updated whilst the DUS is running, how do we handle those cases, have we accounted for that in our script?
> 
>     * If the script fails halfway, how do we ensure that we can run the DUS again once we fix it, taking into consideration that we've also started syncing data.

Aside from data correctness, the script runs in a transaction, doesn't it? So either all rows are copied over or none of it is. If the script fails, it means that the destination tables won't contain the data (except the data that was synced by the web app in the meantime) so you can just re-run it after fixing the bugs.

>     * Do we want to pull in the original `created_at` and `updated_at` values from the `User` table or do we want to set new ones for when the migration takes place?

I second @msarit, I would only copy over the `.updated_at` from the the `users` table. `created_at` should basically be the same as `updated_at` during migration.

>       Maybe we can leverage these timestamps to migrate the data between the first migration and sync in Option 1 or to determine if we have newly updated rows that we missed in Option 2.

Given the script is smart enough to update existing rows, I don't think this matters that much. Worst case scenario you overwrite a row with the same exact data.

File: db/schema.rb (hunk 5 above)
Comment (msarit) [reply db/schema.rb:1379]: Nope, it's not 😟

File: user_settings.sql (hunk 6 above)
Comment (msarit) [reply user_settings.sql:7]: I did this check for `users_notification_settings`; I will double check for `users_settings`.

**Comment by Ridhwana** [comment]
_Really really rough draft of some of the things that the tests will cover (pairing session):_ 

# context "migrating data"
  # it "sets the expected number of records"
  # context "migrates the correct records"
    # it "sets the correct User records"
    # it "sets the correct Profile records"
  # it "sets the correct types" 
  # it "casts data (display_email_on_profile) to the correct types(boolean)"
  # it "sets a fallback value for values that are null"
  # context "for enum settings"
    # it "assigns the correct mapping for enum settings"
    # it "assigns the fallback value when there value passed does not have an enum defined"
  # it "the updated_at and created_at timestamps are more current than the original values"

#  context the user id exists in the user_settings table and user table 
# it replaces the user_settings values with values from the user table

File: db/schema.rb (hunk 5 above)
Comment (rhymes) [reply db/schema.rb:1425]: this should be re-added in place

File: user_settings.sql [hunk 7]
```diff
@@ -0,0 +1,76 @@
+-- This is just a test (in draft) file so that Arit and I can work together using revision control
```
Comment (rhymes) [inline user_settings.sql:1]: this can be removed from the PR now that it's in the data update script

File: users_notification_settings.sql [hunk 8]
```diff
@@ -0,0 +1,42 @@
+-- This is just a test (in draft) file so that Arit and Ridhwana can work together using revision control
```
Comment (rhymes) [inline users_notification_settings.sql:1]: is this going to be moved in a data update script as well?

File: lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
+module DataUpdateScripts
+  class MigrateRelevantFieldsFromUsersToUsersSettings
+    def run
+      ActiveRecord::Base.connection.execute(
+        <<~SQL.squish,
+          BEGIN TRANSACTION;
```
Comment (rhymes) [inline lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb:6]: If you notice in the terminal when you run the spec you'll see warnings like `WARNING:  there is already a transaction in progress`, this is because statements in tests are wrapped in a transaction that's rolled back at the end, thus `BEGIN transaction` is superflous.

That said, this is meant to run on development and production and, if you load the development console and run this script manually with:

```ruby
require_relative 'lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb
DataUpdateScripts::MigrateRelevantFieldsFromUsersToUsersSettings.new.run
```

you'll see how it's actually not wrapped in a transaction.

My suggestion is to use `ActiveRecord::Base.transaction`, this will work both in development:

```ruby
[2] pry(main)> DataUpdateScripts::MigrateRelevantFieldsFromUsersToUsersSettings.new.run
  TRANSACTION (0.4ms)  BEGIN
   (153.2ms)  WITH settings_data AS ...
  TRANSACTION (8.5ms)  COMMIT
```

and in the tests.

Something like:

```ruby
      ActiveRecord::Base.transaction do
        ActiveRecord::Base.connection.execute(...)
      end
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+      expect(Users::Setting.last.display_email_on_profile).to be_in([true, false])
+    end
+
+    it "sets a fallback value for values that are null" do
+      create(:user, display_announcements: nil)
+
+      described_class.new.run
+
+      expect(Users::Setting.last.display_announcements).to eq(true)
```
Comment (rhymes) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:69]: tiny nitpick, this should be `be(true)` like on line 79

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    # Not sure how to test this; when I pass in an invalid font, the spec errors with
+    # "Validation failed: Config font fake_font is not a valid font selection"
+    it "assigns the fallback value when there value passed does not have an enum defined" do
+      create(:user)
+
+      described_class.new.run
+
+      expect(Users::Setting.first.default?).to be(true)
+    end
```
Comment (rhymes) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:90]: You can test this bypassing the validation callbacks:

```ruby
      user = create(:user)
      user.update_columns(config_font: "fakefont")
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  it "the updated_at and created_at timestamps are more current than the original values" do
+    user = create(:user, created_at: 1.minute.ago, updated_at: 1.minute.ago)
+
+    described_class.new.run
+
+    expect(Users::Setting.first.created_at).to be > user.created_at
+    expect(Users::Setting.first.updated_at).to be > user.updated_at
```
Comment (rhymes) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:99]: note: you can save `Users:::Setting.first` to an object and use that as a comparison

```ruby
user_setting = Users::Setting.first
expect(user_setting.created_at.after?(user.created_at)).to be(true)
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  context "when the user id exists in both the users_settings users tables" do
```
Comment (rhymes) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:102]: ❤️

**Review by rhymes — COMMENTED** [review]
Great job @msarit! 

I left a few suggestions in the review :) 

Everything works BTW!!!! 🔥

File: users_notification_settings.sql (hunk 8 above)
Comment (msarit) [reply users_notification_settings.sql:1]: I will check and update 👍🏾

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb (hunk 11 above)
Comment (msarit) [reply spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:90]: Thank you so much for this; it was driving me crazy!

File: lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb (hunk 9 above)
Comment (msarit) [reply lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb:6]: @rhymes Do I also remove the `COMMIT;` at the end of the SQL statement?

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    user = create(:user, created_at: 1.minute.ago, updated_at: 1.minute.ago)
+
+    described_class.new.run
+
+    expect(users_setting.created_at).to be > user.created_at
+    expect(users_setting.updated_at).to be > user.updated_at
+  end
+
+  context "when the user id exists in both the users_settings users tables" do
+    it "replaces the users_setting values with values from the user table" do
```
Comment (msarit) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:105]: @rhymes I'm having issues with this last spec 😓 

Basically, `line 114` doesn't seem to be updating the `display_announcements` setting for the `user`, because the `expect` on `line 120` keeps failing (returning `true` instead `false`).

I described the issue in [this video](https://www.loom.com/share/77b0815ecb1a4e4d96c8a13e28c990d0).

File: lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb (hunk 9 above)
Comment (rhymes) [reply lib/data_update_scripts/20210423155327_migrate_relevant_fields_from_users_to_users_settings.rb]: @msarit Yes, thank you!

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+      expect(users_setting.user_id).to eq(user_id)
+      expect(users_setting.display_announcements).to be(true)
+
+      user.update_columns(display_announcements: false)
+      user.reload
+
+      described_class.new.run
+
+      expect(users_setting.user_id).to eq(user_id)
+      expect(users_setting.display_announcements).to be(false)
```
Comment (rhymes) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:120]: `user.reload` reloads the `User` object, but you're interested in the `User::Settings` value, you have two options:

1. add `user_setting.reload` after `described_class.new.run` so it reloads the value from the DB, which now shall be `false`
2. change the expectation to `expect(user.setting.display_announcements).to be(false)`, this is because `.setting` loads the correct "setting" object from the DB. `User` and `Users::Setting` are related to each other through `user.setting`

Let me know if any of this isn't clear :)

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      user.update_columns(config_font: "fake_font")
+      user.reload
+
+      described_class.new.run
+
+      expect(users_setting.default?).to be(true)
+    end
+  end
+
+  it "the updated_at and created_at timestamps are more current than the original values" do
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:95]: _Quick, non-blocking nitpick_:
```suggestion
  it "assigns updated_at and created_at timestamps that are more current than the original values" do
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    user = create(:user, created_at: 1.minute.ago, updated_at: 1.minute.ago)
+
+    described_class.new.run
+
+    expect(users_setting.created_at).to be > user.created_at
+    expect(users_setting.updated_at).to be > user.updated_at
+  end
+
+  context "when the user id exists in both the users_settings users tables" do
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:104]: _Quick, non-blocking nitpick_:
```suggestion
  context "when the user id exists in both the users_settings and users tables" do
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+    it "replaces the users_settings values with values from the user table" do
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_settings_spec.rb:105]: _Quick, non-blocking nitpick_:
```suggestion
    it "replaces the users_settings values with values from the users table" do
```

File: spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_notification_settings_spec.rb [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    it "sets a fallback value for values that are null" do
+      create(:user, email_newsletter: nil)
+
+      described_class.new.run
+
+      expect(users_notification_setting.email_newsletter).to be(false)
+    end
+  end
+
+  it "the updated_at and created_at timestamps are more current than the original values" do
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/migrate_relevant_fields_from_users_to_users_notification_settings_spec.rb:48]: _Quick, non-blocking nitpick_:
```suggestion
  it "assigns updated_at and created_at timestamps that are more current than the original values" do
```

**Review by juliannatetreault — COMMENTED** [review]
Let a few non-blocking comments, otherwise, this is looking great so far! ✨

**Review by rhymes — APPROVED** [review]
Yes! 🔥 🔥 

Awesome work! :) 

I think once @juliannatetreault suggestions are applied this is ready to be merged 🥳

**Comment by Ridhwana** [comment]
> Yes! 🔥 🔥
> Awesome work! :)
> I think once @juliannatetreault suggestions are applied this is ready to be merged 🥳

@rhymes just a note in case you missed the comment in the description. 

As per your comments below and the example of the profile migration, we're going to implement the work in the ffg order: 

> the more I read your proposal, the more I think we should do something similar to the profile migration, in which the code to sync is put in place before the migration script. This is the scenario I imagine:
> 
> 1. the sync code is deployed
> 2. users objects have .save called upon them
> 3. this saves the data in users and syncs with the new tables using this migration script
> 4. we deploy a data update script to sync all remaining users
> 5. we then start migrating the destination columns from users to users_settings, so that when a profile is saved, the data ends in the latter table

 We're hoping to get the sync code working by next week and then we can deploy this PR shortly thereafter.

**Review by juliannatetreault — APPROVED** [review]
I tested this locally and it looked good to me! 🚀
